### PR TITLE
Models Genesis Transforms

### LINKED ISSUE #98: [FeatureRequest] Add Equivalent to Models Genesis Transform
add equivalent to transforms described in models genesis https://arxiv.org/abs/1908.06912

### PR DESCRIPTION
**Pull request by weningerleon**
## Short Description
- Implementation of Models Genesis Transforms in the rising framework, should close #98 
- Small bug fix for Mirror (before, the a probability of the flip was mentioned in the documentation, but not implemented)

What do you think of this proposition of the models genesis transforms? What kind of tests do you think are necessary?

### CODE REVIEW

File: rising/utils/torchinterp1d.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+import torch
+import contextlib
+SEARCHSORTED_AVAILABLE = True
+try:
+    from torchsearchsorted import searchsorted
+except ImportError:
+    SEARCHSORTED_AVAILABLE = False
+
+
+class Interp1d(torch.autograd.Function):
```
Comment (justusschock) [inline rising/utils/torchinterp1d.py:43]: looking at this repo, we may want to reimplement this part in PyTorch's Python interface (which shouldn't take so long, I'll maybe have a look this weekend).
We would prefer not to have external dependencies :)

File: CODEOWNERS [hunk 2]
```diff
@@ -32,15 +32,18 @@
 /rising/transforms/compose.py @mibaumgartner
 /rising/transforms/crop.py @mibaumgartner
 /rising/transforms/format.py @mibaumgartner
-/rising/transforms/intensity.py @justusschock @mibaumgartner
+/rising/transforms/intensity.py @justusschock @mibaumgartner @weningerleon
 /rising/transforms/kernel.py @mibaumgartner
 /rising/transforms/spatial.py @justusschock @mibaumgartner
 /rising/transforms/utility.py @justusschock @mibaumgartner
+/rising/transforms/painting.py @weningerleon
```
Comment (justusschock) [inline CODEOWNERS:39]: The CODEOWNERS file is basically not for code ownership (name is a bit misleading), but for automated PR review requests. This means if you register yourself here as an owner, you probably will get automated requests to review PRs for these files.

The actual code ownership is handled based on commits and the associated accounts.

File: rising/transforms/functional/painting.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+def random_outpainting(data: torch.Tensor, maxv: float = 1.0, minv: float = 0.0) -> torch.Tensor:
+
+    batch_size, channels, img_rows, img_cols, img_deps = data.size()
+
+    out = torch.rand(data.size()) * (maxv - minv) + minv
+
+    block_size_x = torch.randint(5*img_rows // 7, 6*img_rows // 7, (1,))
+    block_size_y = torch.randint(5*img_rows // 7, 6*img_rows // 7, (1,))
+    block_size_z = torch.randint(5*img_rows // 7, 6*img_rows // 7, (1,))
```
Comment (p-wein) [inline rising/transforms/functional/painting.py:69]:
```suggestion
    block_size_y = torch.randint(5*img_cols // 7, 6*img_cols // 7, (1,))
    block_size_z = torch.randint(5*img_deps // 7, 6*img_deps // 7, (1,))
```
I think x,y,z should depend on img_rows, img_cols, img_deps respectively. (not all on img_rows)

File: rising/transforms/functional/painting.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+                            z:z + block_size_z,
+                        ]
+            idx = torch.randperm(window.numel())
+            window = window.view(-1)[idx].view(window.size())
+
+            data[b, c, x:x + block_size_x,
+                          y:y + block_size_y,
+                          z:z + block_size_z] = window
+
+    return data
```
Comment (p-wein) [inline rising/transforms/functional/painting.py:33]:
```suggestion
def local_pixel_shuffle(data: torch.Tensor, n: int = -1, block_size: tuple=(0,0,0), rel_block_size: float = 0.1) -> torch.Tensor:

    batch_size, channels, img_rows, img_cols, img_deps = data.size()

    if n < 0:
        n = int(1000*channels) # changes ~ 12.5% of voxels
    for b in range(batch_size):
        for _ in range(n):
            c = torch.randint(0,channels-1, (1,))
            
            (block_size_x, block_size_y, block_size_z)  = (torch.tensor([size]) for size in block_size)
            
            if rel_block_size > 0:
                block_size_x = torch.randint(2, int(img_rows * rel_block_size), (1,))
                block_size_y = torch.randint(2, int(img_cols * rel_block_size), (1,))
                block_size_z = torch.randint(2, int(img_deps * rel_block_size), (1,))

            x = torch.randint(0, int(img_rows - block_size_x), (1,))
            y = torch.randint(0, int(img_cols - block_size_y), (1,))
            z = torch.randint(0, int(img_deps - block_size_z), (1,))

            window = data[b, c, x:x + block_size_x,
                            y:y + block_size_y,
                            z:z + block_size_z,
                        ]
            idx = torch.randperm(window.numel())
            window = window.view(-1)[idx].view(window.size())

            data[b, c, x:x + block_size_x,
                          y:y + block_size_y,
                          z:z + block_size_z] = window

    return data
```

Suggestion: Give user the option to define patch sizes themselves, either as absolute sizes or relative to the image size. My Implementation overwrites absolute sizes with relative sizes, if the relative size is > 0 (default = 0.1 to keep default behavior).

File: rising/transforms/functional/painting.py (hunk 3 above)
Comment (weningerleon) [reply rising/transforms/functional/painting.py]: Thanks for the catch!

File: rising/transforms/functional/painting.py (hunk 4 above)
Comment (weningerleon) [reply rising/transforms/functional/painting.py:33]: I think this is a good idea. However, for users, the settings will be difficult to understand -> it needs to be clearer to the user how this works. E.g., at the moment, if the user sets block_size, but the rel_block_size is left to default, block_size will be ignored.
E.g., at least some comments, and perhaps some improvements in the input handling would be great

File: rising/utils/torchinterp1d.py (hunk 1 above)
Comment (weningerleon) [reply rising/utils/torchinterp1d.py:36]: With the new torch.searchsorted function, we don't need the external dependency any more :)

**Review by justusschock — APPROVED** [review]
cc @mibaumgartner  for review

File: rising/utils/torchinterp1d.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+"""
+import torch
+import contextlib
+
+class Interp1d(torch.autograd.Function):
+    def __call__(self, x, y, xnew, out=None):
+        return self.forward(x, y, xnew, out)
```
Comment (justusschock) [inline rising/utils/torchinterp1d.py:38]:
```suggestion
```

File: rising/transforms/functional/intensity.py [hunk 6]
```diff
@@ -227,3 +229,39 @@ def scale_by_value(data: torch.Tensor, value: float,
...
+
+    points = (1-t*t*t)*p0 + 3*(1-t)*(1-t)*t*p1 + 3*(1-t)*t*t*p2 + t*t*t*p3
+
+    # scaling according to maxv,minv
+    points = points*(maxv-minv) + minv
+
+    xvals = points[:,0]
+    yvals = points[:,1]
+
+    out_flat = Interp1d()(xvals, yvals, data.view(-1))
```
Comment (justusschock) [inline rising/transforms/functional/intensity.py:251]:
```suggestion
    out_flat = Interp1d.apply(xvals, yvals, data.view(-1))
```

File: rising/utils/torchinterp1d.py [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+
+    def forward(ctx, x, y, xnew, out=None):
```
Comment (justusschock) [inline rising/utils/torchinterp1d.py:40]:
```suggestion
    @staticmethod
    def forward(ctx, x, y, xnew, out=None):
```

**Review by justusschock — APPROVED** [review]

**Pull request merged**